integration/spark-common/src/main/java/org/apache/carbondata/spark/load/CarbonLoaderUtil.java

```diff
@@ -312,18 +313,28 @@ public static boolean recordLoadMetadata(LoadMetadataDetails newMetaEntry,
           }
           if (insertOverwrite) {
             for (LoadMetadataDetails entry : listOfLoadFolderDetails) {
-              entry.setLoadStatus(CarbonCommonConstants.MARKED_FOR_DELETE);
+              if (!entry.getLoadStatus().equals(LoadStatusType.INSERT_OVERWRITE.getMessage())) {
+                entry.setLoadStatus(CarbonCommonConstants.MARKED_FOR_DELETE);
+                // For insert overwrite, we will delete the old segment folder immediately
+                // So collect the old segments here
+                String path = carbonTablePath.getCarbonDataDirectoryPath("0", entry.getLoadName());
+                staleFolders.add(FileFactory.getCarbonFile(path));
+              }
             }
           }
           listOfLoadFolderDetails.set(indexToOverwriteNewMetaEntry, newMetaEntry);
         }
         SegmentStatusManager.writeLoadDetailsIntoFile(tableStatusPath, listOfLoadFolderDetails
             .toArray(new LoadMetadataDetails[listOfLoadFolderDetails.size()]));
+        // Delete all old stale segment folders
+        for (CarbonFile staleFolder : staleFolders) {
+          CarbonUtil.deleteFoldersAndFiles(staleFolder);
+        }
```

Comment:
but it seems that we delete the folder without checking the operation status here，different from the behavior in another changes. Am I missing something?